fabtests/common/shared.c

```diff
@@ -1920,17 +1920,50 @@ ssize_t ft_tx(struct fid_ep *ep, fi_addr_t fi_addr, size_t size, void *ctx)
 
 ssize_t ft_post_inject(struct fid_ep *ep, fi_addr_t fi_addr, size_t size)
 {
-	if (hints->caps & FI_TAGGED) {
-		FT_POST(fi_tinject, ft_progress, txcq, tx_seq, &tx_cq_cntr,
-			"inject", ep, tx_buf, size + ft_tx_prefix_size(),
-			fi_addr, tx_seq);
+	if (ft_check_opts(FT_OPT_ENABLE_HMEM)) {
+		const struct iovec iov = {
+			.iov_base = tx_buf,
+			.iov_len = size + ft_tx_prefix_size(),
+		};
+
+		if (hints->caps & FI_TAGGED) {
+			const struct fi_msg_tagged msg_tagged = {
+				.msg_iov = &iov,
+				.desc = &mr_desc,
+				.iov_count = 1,
+				.addr = fi_addr,
+				.tag = tx_seq,
+			};
+
+			FT_POST(fi_tsendmsg, ft_progress, txcq, tx_seq,
+				&tx_cq_cntr, "fi_tsendmsg", ep, &msg_tagged,
+				FI_INJECT);
+		} else {
+			const struct fi_msg msg = {
+				.msg_iov = &iov,
+				.desc = &mr_desc,
+				.iov_count = 1,
+				.addr = fi_addr,
+			};
+
+			FT_POST(fi_sendmsg, ft_progress, txcq, tx_seq,
+				&tx_cq_cntr, "fi_sendmsg", ep, &msg,
+				FI_INJECT);
+		}
```

Comment:
This is not the same behavior as the current code.  fi_sendmsg(... FI_INJECT) is not the same as fi_inject().

The existing code must be usable as-is, even if FI_HMEM is enabled.  If it is not, then the provider has broken the API contract and will not work with existing applications.

Reply:
Even though fi_sendmsg(... FI_INJECT) and fi_inject() are not the same, this still works since tx_cq_cntr will not be incremented for fi_sendmsg(... FI_INJECT).

Would you rather have a ft_sendmsg_inject() be defined and have the benchmark apps call it if HMEM is enabled instead of ft_inject()?

Reply:
No - I'm saying that there should be no reason for this change.  If the existing code does not work with a provider, then it's the provider that is broken, not the test.

The fi_inject() call is usable even if the provider has set FI_MR_LOCAL.  There's no indication that setting FI_MR_HMEM changes anything, or that fi_inject is restricted to only host memory.

It may be that the API needs to be updated to address this.  But for now, I believe if a provider cannot support fi_inject() on device memory, then it needs to set the inject_size to 0.

Reply:
I think I follow the conversation enough to agree with @shefty on this. FI_HMEM as a capability shouldn't influence code changes beyond what is documented in the man pages. 

> It may be that the API needs to be updated to address this. But for now, I believe if a provider cannot support fi_inject() on device memory, then it needs to set the inject_size to 0.

I cannot think of a case where a provider couldn't support the inject case. However, if it couldn't -- then having the provider set the inject_size to 0 falls in line with other guidance that has been suggested in the past. I don't like it from a performance perspective, but it's not as if I see another alternative at the moment.

Reply:
@jswaro - I believe the issue is that the provider does not know where the memory resides and needs to discover it.  fi_inject() does not provide any mechanism to convey that information.  This comes across as an area that we need further definition around fi_inject() and FI_HMEM, possibly considering some extension, or at least guidance to apps to replace the use of fi_inject() when FI_MR_HMEM is specified.

Reply:
Is this a case where FI_MR_HMEM wasn't returned and should have been? 

fi_inject doesn't provide a desc field for the memory descriptor, so what your saying makes sense. So in that vein, fi_inject cannot be used with FI_MR_HMEM until transparent memory detection is provided (Ian was working on that but I believe we agreed to pull it from the verbs provider until a hook provider could be implemented), or until the fi_inject definition is changed. 

Changing fi_inject at this point seems like a bad idea. Too many people use the existing definition and the API was designed in mind with main memory, not device memory. Are you suggesting a new API, or an alternative using fi_sendmsg instead? I'm not sure I follow what you mean by extension.

Reply:
I can think of several options.

1 Do nothing.  The provider reports a usable inject size.  fi_inject() would need to determine where a buffer is located if FI_HMEM is enabled.  This would cripple the use of fi_inject() as a performance path when FI_HMEM is set.  The guidance to developers in this case would be to replace fi_inject() with fi_sendmsg().
2 Do nothing.  The provider reports an inject size of 0.
3 Limit fi_inject() to system memory only.  Device memory must use a separate path for inject transfers.
4 As a follow on to 3, introduce a new call fi_hmem_inject() that takes additional arguments to identify where a buffer is located.

Reply:
Option 1 doesn't seem viable. We would force providers to implement device memory detection or require the use of a utility provider to support basic functionality with FI_HMEM. I do not think we can use a utility provider to detect device memory for fi_inject, because the function definition lacks the fields required to convey the necessary information. 

Option 2 is straight forward. Application developers should have taken the inject size into account when they did initialization, but it's a pretty nasty blow to performance. Ideally, MPI wouldn't want to use fi_send, when fi_inject should be available. However.... not much to be done about that. 

Option 3. I think that would be a consistent response. The API doesn't provide for the use case with FI_HMEM, and then behavior would be consistent between providers without requiring a utility layer. Even with a hook provider, fi_inject could not be provided with the information needed. 

Option 4: It's a possibility. I wouldn't want to go with 'fi_hmem_inject' because then maybe we'll need a 'fi_xyz_inject' later for some other feature/functionality. Something more generic perhaps... or just something with the descriptor in the parameters so that core providers can figure out what to do with it.

Reply:
Option 2 is the current option that's available.  And I agree, both 1 & 2 have crippled the use of inject.  We need something before the next release.

The intent of option 4 was to allow additional arguments that can be used to identify the buffer.  As an example:
```
fi_hmem_inject(struct fid_ep *ep, const void *buf, size_t len,
    enum fi_hmem_iface iface, uint64_t device,
    fi_addr_t dest_addr);
```
Anything more generic and we're basically at fi_sendmsg().  Inject calls do not require memory registration (one of the advantages of using inject), so we can't rely on having a memory descriptor.

Option 5: Add a new hmem_inject_size attribute, so the provider can specify different inject sizes.  This fixes the current situation, but doesn't necessarily make it any easier for the provider to have separate fast paths for system versus device memory.

Reply:
Coming late to the party....

As already stated, fi_inject() does not provide the HMEM iface type required by a provider to properly use device memory with fi_inject(). Having the provider discover the HMEM iface type during an inject call would just hurt performance and should be avoided. In addition, have a provider being forced to report an inject_size of zero if HMEM is enabled seems very heavy handed and I think we should avoid this.

> Option 5: Add a new hmem_inject_size attribute, so the provider can specify different inject sizes. This fixes the current situation, but doesn't necessarily make it any easier for the provider to have separate fast paths for system versus device memory.

An hmem_inject_size attribute on its own doesn't seem to fix the current situation. For example, if this value is non-zero, can I use fi_inject() or not?

I think it would make sense to leave fi_inject() as system memory only and introduce the new fi_hmem_inject() API. In addition, we then could depreciate fi_inject().

Reply:
If we went with option 4 or option 5, my only nit is the name of the function. Admittedly, I don't have a better suggestion though. My nit stems from the fact that the new function is similar in nature to fi_sendmsg -- where fi_inject essentially becomes a specialized call of the new function(fi_hmem_inject) where the iface and device are assumed to be system memory. The new function requires the device and iface to be explicitly provided. So ... it is a function that deals with HMEM... true, but we didn't implement a fi_hmem_sendmsg... so it seems to break naming convention in a odd way. 

Again, I don't have a suggestion. I am only trying to convey that the name feels odd compared to other transmit functions we've defined so far.

Reply:
Another observation is that we probably need some field in the fi_info (hmem_inject_size) to convey whether we can provide inject space in the provider for device memory. In some cases, such as with GPUs, we've found that if the buffer belongs to another device, device libraries may assume it belongs to main memory and attempt to do a copy from the buffer to the driver/device. This is something we've seen with the latest Mellanox OFED 5.1, where using GPU memory caused an application to segfault through a feature in the library that enabled a form of coalescing. @iziemba  can provide more details as he was working that particular issue. 

If we provide the additional field, then it would be easy enough to convey to applications whether the provider can inject directly to hardware when using device buffers. Then, implementation of that can be left up to the providers. 

Seems to me, we need another function to support the fi_inject case with device memory, and we probably need a value to determine what the inject length is -- which can also tell the application whether the new function is supported by a provider if the value is 0.

Reply:
So it seems like we have a HMEM inject issues which we should resolve in the emailing list or an issue. In the meantime, what should be done to land this PR? If device memory is requested, should I have the benchmark apps avoid all inject calls?

Reply:
No - the API as defined today, fi_inject is usable with both system and device memory.  The test is fine without any change in this area.  If it does not work with a provider that is a provider bug, not a problem in the test.

The test is correctly pointing out that there is an issue here.  Masking the issue by changing the test should not be the fix.

Reply:
Today to support fi_inject with FI_HMEM today would result in the HMEM iface type being queried on every call fi_inject. Is that the correct solution?

Reply:
It is an API defined solution.  The provider can also disable inject support when FI_HMEM is requested.

It would have been great to have caught this before we shipped a release version of libfabric with FI_HMEM defined.  But we didn't.  An application coding to libfabric v1.11 can pass in device memory directly into the fi_inject() call.  It doesn't really matter how slow or inefficient that is.  Does it work?

Reply:
>  If it does not work with a provider that is a provider bug, not a problem in the test.

We discussed this in the initial proposal. Initially, it was suggested that the provider could detect whether memory was device or system within the provider, but that idea was rejected in favor of providing that type of detection with a utility provider. Using a utility provider to detect memory type won't work with fi_inject case because the API lacks the information necessary to communicate the domain (whether system or device) of the memory to the provider at the time of the call. 

In my opinion, we should provide the capability to detect the type of memory within the provider, or disable the inject interface when FI_HMEM is requested. If a provider is going to provide the implementation of memory type detection within the provider, the FI_MR_HMEM could be excluded from the mr_mode bits. If the provider does not provide memory type detection, then FI_MR_HMEM must be in the mr_mode bits. 

fi_inject lacks the capability to provide the underlying implementation with any details about the memory type of the local buffer. If FI_MR_HMEM is set, then fi_inject cannot be used without the provider performing memory type detection on behalf of middleware or the application writer. It seems to me that we should provide guidance to provider writers to set the inject size to 0, or provide the capability to perform memory type detection when FI_MR_HMEM is set. 

It doesn't seem to me that an application can choose to be unaware of memory type if FI_MR_HMEM is set. The man pages state that applications must register device buffers through the fi_mr_regattr call when FI_MR_HMEM is set, and the only way they know how to make that decision is by being memory type aware in the first place. So to support middle-ware that is not memory type aware, provider maintainers need to perform memory type detection on behalf of the application, or disable the inject interface by setting inject size to 0.

In my head, this looks like we would potentially return 3 different variants of a provider depending on input hints. 

-  an fi_info structure without FI_HMEM
-  an fi_info structure with FI_HMEM but without FI_MR_HMEM (memory type detection performed by provider, kernel, or hardware)
-  an fi_info structure with FI_HMEM and FI_MR_HMEM (memory type detection is not performed by libfabric or underlying layers. fi_inject interface is disabled if inject_size is set to 0, otherwise it is set to some positive value if the kernel or hardware can provide memory type detection transparently and we choose to ignore the FI_MR_HMEM requirement for the fi_inject interface).

Reply:
There is technically no need for the provider to detect the memory type with fi_inject.  They can simply use memcpy to copy the buffer into a pre-registered region or the outbound transmit queue.  Sure, memcpy may be slower than an alternative copy, but it's going to be faster than the application needing to register the buffer and going through a non-inject path.  Additionally, it's not clear than an app will ever want to use inject with a device buffer anyway.  Most device buffers tend to be for large transfers, not small control messages that inject targets.

There's a good chance we're overthinking this and trying to design for a use case that is unlikely to occur in practice.

Reply:
> They can simply use memcpy to copy the buffer into a pre-registered region or the outbound transmit queue. Sure, memcpy may be slower than an alternative copy, but it's going to be faster than the application needing to register the buffer and going through a non-inject path.

That's the problem though. A provider cannot use memcpy from the device to a bounce/inject buffer unless there is specific kernel support in place for example. Use of memcpy without kernel support would result in a segmentation fault.

Reply:
FI_HMEM already requires specialized, out of tree kernel support by both the GPU and NIC for any of this to work.  Do the GPU drivers not already support this behavior?  I thought they did.

A util provider can still handle this.  It just needs to copy into a host buffer and pass that to the underlying provider's inject call.

Reply:
> A util provider can still handle this. It just needs to copy into a host buffer and pass that to the underlying provider's inject call.

Sure, that could work. 

> FI_HMEM already requires specialized, out of tree kernel support by both the GPU and NIC for any of this to work. Do the GPU drivers not already support this behavior?

Well, GPUs like Nvidia's require nv_peer_mem, that much is true. They do require kernel support provide the bridge between the device and the NIC. What I was referring to was support of memory movement between device and host transparently and facilitated by newer kernel features such as HMM.  Without HMM or features like HMM, applications would still need to make device specific calls to copy memory. 

But from user space, you can't call memcpy on the buffer without something like HMM. You need to call the CudaMemcpy -- or equivalent -- in order to transfer memory from the device to the host's bounce buffer in order to use inject. The other option as described above would be to use a utility provider to do the detection and use a bounce buffer in the utility provider to continue to use the core provider's fi_inject function.

Reply:
@j-xiong @aingerson  - Can you confirm that using memcpy on a GPU buffer requires HMM, or can the driver mmap support handle this?

Reply:
The GPU buffer pointer itself is not an mmap-ed address. To use it directly in memcpy() requires HMM.  If mmap() is called explicitly to created a different address for the buffer, then that can be used in memcpy. Notice that  the driver may migrate the buffer when handling the mapping.

Reply:
Returning to the original problem, for this patch series, please revert the above changes and use the existing fi_inject calls as-is.